Allow Request's Headers to be created with various objects

jeenalee opened the pull request:
<!-- Please describe your changes on the following line: -->

While Headers could be constructed correctly with an array or
object (open ended dictionary/MozMap), Request's Headers failed to be
created with non-Headers object (such as array or open ended
dictionary/MozMap).

Before, Request's Headers could be filled with only a Headers object in
Step 28. This has been expanded to accommodate array and open ended
dictionary.

Step 29 empties the Request's Headers list after it had been filled in
Step 28, thus resulting in an empty Headers object when it shouldn't
be. This step has been removed with a comment in this commit.

If a RequestInit Headers is _not_ given, but a RequestInfo Headers is
given, RequestInfo Headers should be used to construct Request
Headers. That step has been added after Step 31.

Corresponding wpt result is updated in this commit.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #13758 (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13802)

<!-- Reviewable:end -->

jeenalee committed:
Allow Request's Headers to be created with various objects

While Headers could be constructed correctly with an array or
object (open ended dictionary/MozMap), Request's Headers failed to be
created with non-Headers object (such as array or open ended
dictionary/MozMap).

Before, Request's Headers could be filled with only a Headers object in
Step 28. This has been expanded to accommodate array and open ended
dictionary.

Step 29 empties the Request's Headers list after it had been filled in
Step 28, thus resulting in an empty Headers object when it shouldn't
be. This step has been removed with a comment in this commit.

If a RequestInit Headers is *not* given, but a RequestInfo Headers is
given, RequestInfo Headers should be used to construct Request
Headers. That step has been added after Step 31.

Corresponding wpt result is updated in this commit.

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/dom/request.rs
 * @kichjang: components/script/dom/request.rs

emilio commented on components/script/dom/request.rs:
```diff
@@ -308,21 +308,26 @@ impl Request {
         // Step 29
-        r.Headers().empty_header_list();
+        // When r.Headers()'s header list is emptied,
+        // r.Headers that was filled in Step 28 becomes empty.
```
This comment seems confusing to me, could you explain or rephrase it? It seems we can't empty these because otherwise we undo the stuff done above?

If that's the case, we should be doing a deep copy in step 27 then instead.

emilio replied:
So that backtrace can come from anywhere in the constructor, so it's hard to diagnose based on that.

But a borrow error indicates that you're reading from something while writing to it in general. Something like filling r.headers() with themselves, or holding a borrow to somewhere inside the Headers object can be potential problems.

jeenalee replied:
Hmm. I realized I wasn't making a deep copy. I had to do `r.Headers.r().clone()` in order to actually make a deep copy.

With this approach though, because `r.Headers()` is a `Root<T>`, it becomes difficult to make a mutable reference to `r.Headers()`. Without the mutable reference, we cannot set `r.Headers()` to be the deep copied headers in Step 27.

I want to suggest we leave the comment more descriptive (as you wrote above), but not make a deep copy in Step 27.

emilio commented on components/script/dom/request.rs:
```diff
@@ -341,7 +346,19 @@ impl Request {
+        // "associated headers list". If an init headers is not given,
+        // but an input with headers is given, set request's
+        // headers as the input's Headers.
+        if let None = init.headers {
```
nit: if `init.headers.is_none`, though this can probably be more legible with a match statement combined with the above `if let`.

emilio commented on components/script/dom/request.rs:
```diff
@@ -308,21 +308,26 @@ impl Request {
-            headers_copy = input_request.Headers();
-        }
-
         // Step 28
```
The step numbers seem to have changed (though that's probably my fault!), feel free to update them if you feel like it :).

jeenalee replied:
I'll leave them as it is for now! It's probably more straightforward to fix them when we update Request Constructor to match the spec :)

emilio reviewed (commented):
Sorry for the _huge_ delay @jeenalee :(. Feel free to ping me more often if you see this happens again.

jeenalee commented:
@emilio Thanks Emilio! I realized that the Fetch spec has been updated, but our request constructor has not yet. The changes in this PR were based on the older Fetch spec version. Will it make more sense to close this PR, and update the request constructor to match the current Fetch spec first?

jdm commented:
Updating the numbers in a separate commit in the same PR is fine; or doing it separately after. I doubt it's necessary to close this one though.

jeenalee committed:
Address review comments

jeenalee committed:
Remove unused use statement

jeenalee commented on components/script/dom/request.rs:
```diff
@@ -308,21 +309,31 @@ impl Request {
+        // `r.Headers()` is a `Root<T>`, and therefore it is difficult
+        // to obtain a mutable reference to `r.Headers()`. Without the
+        // mutable reference, we cannot mutate `r.Headers()` to be the
+        // deep copied headers in Step 27.
```
Good idea! Hmm, though while it compiles, more tests fail with that change. I'm not sure why.

emilio replied:
Hmmm... So it seems to me we'd need to add a method to `Request` that creates a new `Request` with a clone of the inner hyper list, is that right?

I'm totally fine doing it as a followup if @jdm agrees.

jeenalee commented:
I thought this would be a small change until the request Constructor gets updated to match the Fetch spec. While this patch isn't the perfect fix, I feel like we're starting to go down the rabbit hole... I'd say we move forward with the not-perfect fix, or I can close this pull request and focus on updating the constructor!

emilio reviewed (approved):
The rest looks ok to me, thanks for looking into this @jeenalee! :)

jeenalee commented:
Hi! Just checking in with this PR. Please let me know if there's anything I can do :)

jdm commented:
Whoops, I didn't realize this was waiting on my input. I'm not thrilled with either the before or after code (or the spec, for that matter), but I don't have concrete suggestions for how to improve any of those pieces. Let's merge this and move on!

@bors-servo: r=emilio

bors-servo commented:
:pushpin: Commit e821da8 has been approved by `emilio`

<!-- @bors-servo r=emilio e821da81ce89c205a7708e1448c5ca4d1e181b17 -->

bors-servo commented:
:hourglass: Testing commit e821da8 with merge 4fc283e...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/699)

jdm commented:
Those are some very unhappy tests! One clear problem:
```
  ▶ CRASH [expected OK] /fetch/api/request/request-consume.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ DOMRefCell<T> already borrowed: BorrowMutError (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at ../src/libcore/result.rs:837)
  │ stack backtrace:
  │    0:     0x7fb63064eafd - backtrace::backtrace::trace::h16372ee7bf1517e5
  │    1:     0x7fb63064f192 - backtrace::capture::Backtrace::new::hcc43c50c4b11c693
  │    2:     0x7fb62f3b00fd - servo::main::_{{closure}}::h8e9c5971861a4720
  │    3:     0x7fb631015b24 - std::panicking::rust_panic_with_hook
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:452
  │    4:     0x7fb6310159e4 - std::panicking::begin_panic<collections::string::String>
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:413
  │    5:     0x7fb631015909 - std::panicking::begin_panic_fmt
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:397
  │    6:     0x7fb631015897 - std::panicking::rust_begin_panic
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:373
  │    7:     0x7fb6310522fd - core::panicking::panic_fmt
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libcore/panicking.rs:69
  │    8:     0x7fb62f9a5035 - core::result::unwrap_failed::h643f371f836203d8
  │    9:     0x7fb62fb348e6 - _<script..dom..headers..Headers as script..dom..bindings..codegen..Bindings..HeadersBinding..HeadersBinding..HeadersMethods>::Append::h1c59f2b40c66fb7e
  │   10:     0x7fb62fb356ce - script::dom::headers::Headers::fill::hd3324ebf20b36137
  │   11:     0x7fb62fbce43f - script::dom::request::Request::Constructor::h3329b257fb1e6789
  │   12:     0x7fb62f896371 - std::panicking::try::do_call::h6e04d3e7d58d6235
```
I suspect this is the case where headers_copy is both the object being used to fill the new request's headers and the headers being filled - we'll be calling borrow_mut on one and borrow on the other, which won't work.

jeenalee committed:
Fill r's headers with headers_copy when HeadersInit::Headers is given

Instead of filling request's headers whenever a `HeadersInit` is given, this
patch fills request's headers only when `HeadersInit` with a type of
`Headers` is given. Previously, the constructor tried to fill request's headers
with itself, causing Servo to crash.

jeenalee commented:
Oopsies, that's bad! The latest commit should fix that crash.

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit bfd999f with merge 21bc4e0...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/708)

highfive commented:
▶ FAIL [expected PASS] /_mozilla/css/inline_block_opacity_change.html
      └   → /_mozilla/css/inline_block_opacity_change.html aed6845267bba6c52d3aa69c4889449b454d8117
    /_mozilla/css/inline_block_opacity_change_ref.html febec1b4730afab195f280694dad47200b09ea3c
    Testing aed6845267bba6c52d3aa69c4889449b454d8117 == febec1b4730afab195f280694dad47200b09ea3c

emilio commented:
@bors-servo r+

 * #13480

bors-servo commented:
:pushpin: Commit bfd999f has been approved by `emilio`

<!-- @bors-servo r=emilio bfd999f71e198cffb7b1e84d1848c871937e3bd8 -->

KiChjang commented:
@bors-servo retry clean

jdm commented:
@bors-servo: r-
Hold on, I just want to make sure I understand the implications of the latest code change.

jdm commented:
@bors-servo: r+
Ok, I understand how it works.

bors-servo commented:
:pushpin: Commit bfd999f has been approved by `jdm`

<!-- @bors-servo r=jdm bfd999f71e198cffb7b1e84d1848c871937e3bd8 -->

bors-servo commented:
:hourglass: Testing commit bfd999f with merge 9a7559f...

bors-servo commented:
:snowman: The build was interrupted to prioritize another pull request.

bors-servo commented:
:zap: Previous build results for [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4089) are reusable. Rebuilding only [arm32](http://build.servo.org/builders/arm32/builds/4005), [arm64](http://build.servo.org/builders/arm64/builds/4011), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4104), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/747), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/746), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4085), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/770), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/782), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2017)...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/4007), [arm64](http://build.servo.org/builders/arm64/builds/4013), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4106), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/749), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/748), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4089), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4087), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/772), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/784), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2019)
